crypt.go

```diff
@@ -0,0 +1,488 @@
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return xerrors.Errorf("cannot create stderr pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	done := make(chan bool, 2)
+	go func() {
+		rd := bufio.NewScanner(stdout)
+		for rd.Scan() {
+			fmt.Printf("systemd-cryptsetup: %s\n", rd.Text())
+		}
+		done <- true
+	}()
+	go func() {
+		rd := bufio.NewScanner(stderr)
+		for rd.Scan() {
+			fmt.Fprintf(os.Stderr, "systemd-cryptsetup: %s\n", rd.Text())
+		}
+		done <- true
+	}()
+	for i := 0; i < 2; i++ {
+		<-done
+	}
+
+	return wrapExecError(cmd, cmd.Wait())
+}
+
+func askPassword(sourceDevicePath, msg string) (string, error) {
+	cmd := exec.Command(
+		"systemd-ask-password",
+		"--icon", "drive-harddisk",
+		"--id", filepath.Base(os.Args[0])+":"+sourceDevicePath,
+		msg)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return "", wrapExecError(cmd, err)
+	}
+	result, err := out.ReadString('\n')
+	if err != nil {
+		return "", xerrors.Errorf("cannot read result from systemd-ask-password: %w", err)
+	}
+	return strings.TrimRight(result, "\n"), nil
+}
+
+func getPassword(sourceDevicePath, description string, reader io.Reader) (string, error) {
+	if reader != nil {
+		scanner := bufio.NewScanner(reader)
+		switch {
+		case scanner.Scan():
+			return scanner.Text(), nil
+		case scanner.Err() != nil:
+			return "", xerrors.Errorf("cannot obtain %s from scanner: %w", description, scanner.Err())
+		}
+	}
+	return askPassword(sourceDevicePath, "Please enter the "+description+" for disk "+sourceDevicePath+":")
+}
+
+func decodeRecoveryKey(passphrase string) ([]byte, error) {
+	// The recovery key should be provided as 8 groups of 5 base-10 digits, with each 5 digits being converted to a 2-byte number to
+	// make a 16-byte key.
+	var key bytes.Buffer
+	for len(passphrase) > 0 {
+		if len(passphrase) < 5 {
+			return nil, errors.New("incorrectly formatted (insufficient characters)")
+		}
+		x, err := strconv.ParseUint(passphrase[0:5], 10, 16)
+		if err != nil {
+			return nil, errors.New("incorrectly formatted (invalid base-10 number)")
+		}
+		binary.Write(&key, binary.LittleEndian, uint16(x))
+
+		// Move to the next 5 digits
+		passphrase = passphrase[5:]
+		// Permit each set of 5 digits to be separated by '-', but don't allow the recovery key to end or begin with one.
+		if len(passphrase) > 1 && passphrase[0] == '-' {
+			passphrase = passphrase[1:]
+		}
+	}
+	return key.Bytes(), nil
+}
+
+// RecoveryReason indicates the reason that a volume had to be activated with the fallback recovery key instead of the TPM sealed key.
+type RecoveryReason uint8
+
+const (
+	// RecoveryReasonUnexpectedError indicates that a volume had to be activated with the fallback recovery key because an unexpected
+	// error was encountered during activation with the TPM sealed key.
+	RecoveryReasonUnexpectedError RecoveryReason = iota + 1
```

Comment:
using just the Recovery prefix when recovery key is meant is slightly confusing, given we have many "recovery" concepts around already (systems, partitions etc)

Reply:
my suggestions are:

1. s/RecoveryReason/RecoverKeyRecourse/
2. if 1. sounds too weird, it might or might not, then be very explicit with something like s/RecoveryReason/RecoverKeyUsageReason/

Reply:
I think RecoveryKeysUsageReason sounds better, even if it's longer.